### PR TITLE
Peer cache overhaul

### PR DESCRIPTION
**Pull request by kgaillot**
These revisions fix RHBZ#1193499 (the "member weirdness" issue) by overhauling peer cache management. Rather than being managed by each daemon's callback, the caches are managed by libcluster automatically, ensuring that exiting nodes are removed from the cache. This incidentally fixes a number of other minor issues, including potential use-after-free bugs, a potential iterator invalidation bug, and incorrect log messages.

### CODE REVIEW

**Comment by kgaillot** [comment]
Planned testing for these changes:
* CTS with 75 random tests, on both rhel6 (cman) and rhel7 (corosync2) clusters.
* A "member weirdness" reproducer on both cluster types (in a 5-node cluster, remove three nodes then add one back so it reuses a different node ID).

**Comment by kgaillot** [comment]
FYI b85c008 requires f155e78 to work properly, otherwise crmd will have issues. If that's a problem I can merge the two commits or move some of f155e78 into the earlier one.

Also, a very minor note: in mixed-version clusters, "member weirdness" can still happen with pacemakerd (even on nodes with this fix) because the older-version nodes will broadcast process lists that erase the messaging layer flag, making node exit undetectable.

File: mcp/pacemaker.c [hunk 1]
```diff
@@ -621,15 +637,16 @@ update_node_processes(uint32_t id, const char *uname, uint32_t procs)
...
+
         } else {
             crm_trace("Node %s still has process list: %.32x", node->uname, procs);
         }
     }
-
-    if (changed && id == local_nodeid) {
-        update_process_clients(NULL);
-        update_process_peers();
-    }
```
Comment (beekhof) [inline mcp/pacemaker.c]: I prefer the less nested style, but ok

File: lib/cluster/cpg.c [hunk 2]
```diff
@@ -369,26 +369,33 @@ pcmk_cpg_membership(cpg_handle_t handle,
...
-        crm_info("Joined[%d.%d] %s.%u ", counter, i, groupName->value, joined_list[i].nodeid);
+        crm_info("Node %u joined group %s (counter=%d.%d)",
+                 joined_list[i].nodeid, groupName->value, counter, i);
     }
 
     for (i = 0; i < member_list_entries; i++) {
         crm_node_t *peer = crm_get_peer(member_list[i].nodeid, NULL);
 
-        crm_info("Member[%d.%d] %s.%u ", counter, i, groupName->value, member_list[i].nodeid);
+        crm_info("Node %u still member of group %s (peer=%s, counter=%d.%d)",
```
Comment (beekhof) [inline lib/cluster/cpg.c]: Except `member_list` is a superset of `joined_list`.

**Pull request merged**